Formatter/Pool.php

```diff
@@ -29,20 +31,59 @@ class Pool
     protected $defaultFormatter;
 
     /**
+     * NEXT_MAJOR: use LoggerAwareTrait.
+     *
      * @var LoggerInterface
      */
     protected $logger;
 
     /**
-     * @param LoggerInterface|null $logger
-     * @param string|null          $defaultFormatter
+     * @param string|null $defaultFormatter
      */
-    public function __construct(LoggerInterface $logger = null, $defaultFormatter = null)
+    public function __construct($defaultFormatter = null)
     {
-        $this->logger = $logger;
+        if (func_num_args() == 2) {
+            list($logger, $defaultFormatter) = func_get_args();
+            $this->logger = $logger;
+            $this->defaultFormatter = $defaultFormatter;
+        } elseif (func_num_args() == 1) {
+            if ($defaultFormatter instanceof LoggerInterface) {
+                $this->logger = $defaultFormatter;
+            } else {
+                // NEXT_MAJOR: Only keep this block
+                $this->defaultFormatter = $defaultFormatter;
+            }
+        }
 
-        // TODO: This should become a required first parameter when the major version is changed
-        $this->defaultFormatter = $defaultFormatter;
+        // NEXT_MAJOR: keep the else block only
+        if ($this->logger) {
+            @trigger_error(sprintf(
+                'Providing a logger to %s through the constructor is deprecated since 3.x'.
+                ' and we no longer be possible in 4.0'.
+                ' This argument should be provided through the setLogger() method.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
+        } else {
+            $this->logger = new NullLogger();
+        }
+        // NEXT_MAJOR: make defaultFormatter required
+        if (is_null($defaultFormatter)) {
```

Comment:
This issue is still not resolved: if the first and only arg is a logger, default formatter won't be null here, missing the deprecation trigger. Either check the class property, or set the default formatter on line 51 to null.